Add enemy rando

Chris-Is-Awesome opened the pull request:
Draft PR. This is a work in progress and I'll request review once it's done.

The idea for the first version of enemy rando is it will only randomize common enemies and not bosses. I _would_ like the option later to allow randomizing bosses too, but that I imagine will require a lot more work.

This also introduces `Preloader`, which is a utility class to cache any number of objects across scenes.

Chris-Is-Awesome committed:
add preloader

Chris-Is-Awesome committed:
add preloader

roseasromeo commented:
Bosses wise, I think I likely can eventually swap out the entrances to the Silent Servant fights (for the ability upgrades) and the Avarices within themselves (and likely with each other, though that's more logically complicated). I'm not sure the bosses work in each other's arenas--Betty would just fall off Grandma and Frog's domains, Frog couldn't hop, etc.

Let me know what support would be helpful as you work on things (ex., I'm happy to make a toggle option for the in-game menu UI, or investigate our other menu options)

Chris-Is-Awesome committed:
add stack trace to error logs

Chris-Is-Awesome committed:
add stack trace to error logs

Chris-Is-Awesome committed:
cache enemies and add enemy spawning

Chris-Is-Awesome committed:
cache enemies and add enemy spawning

Chris-Is-Awesome committed:
improve bepinex console logging

Chris-Is-Awesome committed:
improve bepinex console logging

Chris-Is-Awesome commented:
> I wonder if it makes sense to put the enemy randomizer into its own mod (to share back with the Death's Door community) and then we can add a dependency on that mod? I think Enemy Randomizer would be fun for folks outside of AP.

That is a good point. Though both enemy rando and item rando (for enemy traps) would use the preload system. So I would need to add another mod, like ModUtils or something that exposes that method for all mods.

Another option is I could have enemy rando still be part of this AP mod, but have it be toggleable with or without AP being connected.

Chris-Is-Awesome committed:
replace enemies

Chris-Is-Awesome committed:
replace enemies

Chris-Is-Awesome commented:
Enemies are now replaced! I have not done a full playtest of each enemy's behavior once spawned, so we'll need to note any weird behavior down.

I chose to not replace the silent servants because I don't know of a good way to trigger the OnDeath cutscene when a different enemy is in their place, so you'd get stuck and miss out on an item. But threat not, they are still in the pool though :)

Also all the Pot Mimic enemies have been added to the pool. I somehow missed them when tallying up all enemies.

Going forward, I would like to add weighting to each enemy, as there are so many mini-bosses/elite enemies compared to common enemies, so I would like for them to be less common. Either that or I set the enemy's health equal to whatever the vanilla enemy's was.

Chris-Is-Awesome committed:
add weights and fix FireMage

Chris-Is-Awesome committed:
add weights and fix FireMage

Chris-Is-Awesome commented:
I've now added weights to the enemies. There are 5 tiers of weights based on enemy difficulty. These are their percent chances of spawning:

Tier 1: 50%
Tier 2: 30%
Tier 3: 15%
Tier 4: 5%
Tier 5: 1%

roseasromeo committed:
add configuration for each enemy type

also tiny typo fixes

Chris-Is-Awesome committed:
fix a couple bugs
- Don't replace Grey Crow boss
- Enemy invisibility trap now properly resets when scene reloads

Chris-Is-Awesome committed:
merge config changes

Chris-Is-Awesome committed:
remove extra foreach for config binding

roseasromeo committed:
patch LevelSpawn.Awake to suppress the Avarice falling animation

Chris-Is-Awesome committed:
Merge pull request #63 from roseasromeo/enemy-rando-prevent-fall

Patch LevelSpawn.Awake to suppress the Avarice falling animation

Chris-Is-Awesome committed:
fix some bugs
- Fixed most null references that would happen during preloading
- Fixed title screen audio continuing to play after preloading
- Fixed issue with player spawn animation being weird after preloading (thanks rose!)
- Fixed issue with boomerang throwers jumping in place
- Removed PlagueKnightSlowFire as they had an issue with their pathfinding reulsting in them getting stuck on walls sometimes

Chris-Is-Awesome commented:
There are still a couple lingering bugs that I need more info on from testing before I can look into a fix for them. These include:
- There was one case where an AudioSource seemed to get duplicated over time (likely from one enemy doing something specific). No idea what caused this.
- My friend had some combat arenas auto-clear without enemies spawning. But I've never had this happen.
- Bats (both white and black) seem to get above the culling plane for the level if they spawn at a high enough Y position. When this happens, the bat can flash rapidly as it's at the edge of the culling range. I might not bother fixing this one, as it's minor.

roseasromeo commented:
What do you think about putting the Enemy Randomizer toggle on the Title Screen options menu? That would let us skip preloading if someone doesn't have it on, and they can quit to menu to turn it on/off, since it only takes effect on scene load anyway.

I can do the implementation of the toggle this evening if you think that strategy makes sense.

Chris-Is-Awesome committed:
add enemy rando toggle to options menu

Chris-Is-Awesome committed:
improve exclusion handlign and exclude more enemies

Chris-Is-Awesome commented:
I improved exclusion handling, so now you can exclude enemies from replacement by:
- Excluding all enemies in a scene
- Excluding enemies with a specified AI_Brain type in a specified scene
- Excluding enemies with a specified parent in a specified scene
- Excluding enemies with a specified position in a specified scene
- Excluding enemies with a specified name in a specified scene

The current exclusions are:
- All bosses and cutscene actors for those bosses
- All slimes in the glass vials in Furnace scenes
- A few missing stragglers (like the extra enemy in Lord of Doors boss fight)

I would like to do one more async before calling it good to merge.

roseasromeo reviewed (approved):
I think this branch contains all the Enemy Rando fixes from the testing in Sept/Oct.

The one remaining loose thread is that Enemy Rando on should be tested in areas with enemy locked doors and make sure that the doors either default open or correctly open, whichever is the behavior you desire.

Chris-Is-Awesome committed:
Merge branch 'main' into enemy-rando

Chris-Is-Awesome committed:
remove unused code

Chris-Is-Awesome committed:
I think this merges rose's changes to enemy-rando

Chris-Is-Awesome committed:
some final fixes/adjustments

- Fixed Lord of Doors ending being impossible (due to both entrance & enemy randos having bugs)
- Fixed additional null ref with entrance rando regarding the Lord of Doors fight
- Update mod version number
- Logger now takes objects as arguments instead of strings
